### DISCUSSION TITLE
Obtaining position data (Line, Column) while parsing

### DISCUSSION THREAD

**Discussion by TheRealGlumfish** [kevinmehall/rust-peg#369] [Q&A] [answered]
Hi, I am building an interpreter and would like to store where my expressions/statements were parsed from on the source code to my AST nodes. Now internally, the parser input must implement the `position_repr` function as part of the `Parse` trait, which is implemented on `str` (which I'm using).

Question is how can I hook into that to obtain line and column data within a rule? Currently only the `position!()` macro exists which returns an offset. Is there any way to obtain the line and column data without having to separately do a pass on the input to count the newlines and convert the offset to Line and Column.

Does a more advanced `position!()`-like macro need to be implemented? If yes I'd be happy to contribute.

**Comment by kevinmehall** [comment, marked as answer]
The [built-in conversion from offset to line/col](https://github.com/kevinmehall/rust-peg/blob/d51f5c49075db902a0666c330e2d72cae4554aab/peg-runtime/str.rs#L35) for error reporting is O(n), so you wouldn't want to use it for every AST node. Usually you'd store the byte offsets or similar in the AST and convert to line and column only when reporting errors or for use in protocols like LSP. A data structure like [codemap](https://github.com/kevinmehall/codemap) may be useful for this, which also includes line number mapping using binary search into an array of line start offsets.

**Reply by TheRealGlumfish** [reply]
Thanks for your helpful answer and for creating `rust-peg`! I'll have a look at the linked crate!
